docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,152 @@
+# Image Promotion
+
+## Abstract
+
+This proposal describes the best practices and defines patterns for promoting Docker
+images between different stages (eg. from "development" to "production").
+This proposal also describes priciples a way how to configure OpenShift to serve as
+a Continuous Delivery tool.
+
+## Constraints and Assumptions
+
+This document describes the following ways to promote images
+
+* Based on human intervention
+* Driven purely by OpenShift
+* Based on external Continuous Delivery tool (Jenkins, go.cd, Snap, ...)
+
+Docker images can be promoted (and the use-cases are valid for):
+
+* Within a single Project
+* Between multiple Projects
+* Between multiple Clusters
+
+The Docker image promotion can be done using the Docker image tags and LABELs.
+You can also use other metadata that OpenShift adds to the Image resource,
+such as annotations or labels, to allow conditional promotion, but this document
+does not describe these advanced scenarios.
+
+Another constraint is that in some cases the image promotion might not be enough,
+and you want to promote a set of images or entire deployment as one single unit
+of promotion.
+
+## Use Cases
+
+### Automatic promotion based on running tests in Jenkins
+
+As a developer of a Rails application, when I update the source code and push it
+to a remote source code repository, I want Jenkins to detect this change and perform a S2I
+build of my application in OpenShift. Depending of the build result, I want Jenkins to
+promote the newly built Docker image by tagging it as "qa-ready" and use OpenShift
+deployments to promote it as "production-ready".
+
+#### Example
+
+As a user I create two DeploymentConfig's:
+
+* "frontend-test"
+* "frontend-production"
+
+The "frontend-test" DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the "qa-ready" ImageStreamTag. This DeploymentConfig also defines the "post" lifecycle
+hook that executes a script in the container deployed and performs additional verification
+and tests.
+
+Example of DeploymentConfig post-lifecycle hook:
+```json
+"strategy": {
+  "type": "Rolling",
+  "rollingParams": {
+    "pre": {
+      "failurePolicy": "Abort",
```

Comment:
what actually happens if a post-deploy hook fails?  do we actually rollback the deployment?  will the deployment then happen again 2 minutes later when the deployment controller again detects a newer image is available than what is deployed?